llvm/include/llvm/CAS/ActionCache.h

```diff
@@ -0,0 +1,100 @@
+//===- llvm/CAS/ActionCache.h -----------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CAS_CASACTIONCACHE_H
+#define LLVM_CAS_CASACTIONCACHE_H
+
+#include "llvm/ADT/Optional.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/CAS/CASReference.h"
+#include "llvm/Support/Error.h"
+
+namespace llvm::cas {
+
+class CASDB;
+class CASID;
+class ObjectProxy;
+
+/// A key for caching an operation.
+/// It is implemented as a bag of bytes and provided convenient constructor
+/// for string/CAS types.
+// TODO: Support CacheKey other than a CASID but rather any array of bytes.
+// To do that, ActionCache need to be able to rehash the key into the index,
+// which then `getOrCompute` method can be used to avoid multiple calls to
+// has function.
+class CacheKey {
+public:
+  StringRef getKey() const { return Key; }
+
+  CacheKey(const CASID &ID);
+  CacheKey(const ObjectProxy &Proxy);
+  CacheKey(const CASDB &CAS, const ObjectRef &Ref);
+
+private:
+  std::string Key;
+};
+
+/// A cache from a key describing an action to the result of doing it.
+///
+/// Actions are expected to be pure (write-once, collision is an error).
+class ActionCache {
+  virtual void anchor();
+
+public:
+  /// Get a previously computed result for \p ActionKey.
+  Optional<ObjectRef> get(const CacheKey &ActionKey) const {
```

Comment:
Your comment above mentions that the intention is to support arbitrary byte strings as keys - what's the plan for *values*? Will they also be arbitrary bytes, or are they required to be CAS objects?

Reply:
TBD but I am leaning towards always requiring CAS objects since it is a more efficient storage and can avoid unnecessary "poison" entries.

Reply:
>  I am leaning towards always requiring CAS objects since it is a more efficient storage

Agreed.

Reply:
SGTM.